File: day_3/part_1.py
BITS = 12


def f(ns):
    delta = 0
    epsilon = 0

    for b in range(BITS):
        foo = sum(((n & (2 ** b)) >> b) for n in ns)

        if foo > len(ns) // 2:
            delta |= 2 ** b
        else:
            epsilon |= 2 ** b

    return delta * epsilon

if __name__ == "__main__":
    with open('input.txt', 'r') as fi:
        foo = [int(i, 2) for i in fi.readlines()]
    print(f(foo))

File: day_3/src/part_1.rs
use std::include_str;


fn main() {
    let bits = 12;

    let ns: Vec<i32> = include_str!("../input.txt")
        .lines()
        .map(|line| i32::from_str_radix(line, 2).unwrap())
        .collect();

    let mut delta: i32 = 0;
    let mut epsilon: i32 = 0;

    for b in 0..bits {
        let thing = 2i32.pow(b);
        let foo: i32 = ns.iter()
            .map(|n| (n & thing) >> b)
            .sum();

        if foo > (ns.len() / 2).try_into().unwrap() {
            delta |= thing;
        } else {
            epsilon |= thing;
        }
    }

    println!("{}", delta * epsilon);

}
